src/index.test.tsx

```diff
@@ -204,3 +204,162 @@ test('changing subscribable instances', async () => {
   await rep1.close();
   await rep2.close();
 });
+
+test('using isEqual', async () => {
+  const {promise, resolve} = resolver();
+
+  const sentinel = Symbol();
+
+  class FakeReplicache implements Subscribable<ReadTransaction> {
+    subscribe<Data>(
+      query: (tx: ReadTransaction) => Promise<Data>,
+      {
+        onData,
+        isEqual = (a, b) => JSON.stringify(a) === JSON.stringify(b),
+      }: {
+        onData: (data: Data) => void;
+        isEqual?: ((a: Data, b: Data) => boolean) | undefined;
+      },
+    ): () => void {
+      const data = query({} as ReadTransaction);
+      let previous: Data | typeof sentinel = sentinel;
+      void data.then(data => {
+        if (previous === sentinel || isEqual(previous, data)) {
+          previous = data;
+          return onData(data);
+        }
+      });
+
+      return () => undefined;
+    }
+  }
+
+  function A({
+    rep,
+    def,
+  }: {
+    rep: FakeReplicache | null | undefined;
+    def: string;
+  }) {
+    const subResult = useSubscribe(
+      rep,
+      async () => {
+        resolve();
+        return 123n;
+      },
+      {
+        isEqual(a, b) {
+          return a === b;
+        },
+        default: def,
+      },
+    );
+    return (
+      <div>
+        {typeof subResult}, {String(subResult)}
+      </div>
+    );
+  }
+
+  const div = document.createElement('div');
+
+  render(<A key="a" rep={null} def="a" />, div);
+  expect(div.textContent).to.equal('string, a');
+
+  render(<A key="b" rep={undefined} def="b" />, div);
+  expect(div.textContent).to.equal('string, b');
+
+  const rep = new FakeReplicache();
+
+  render(<A key="c" rep={rep} def="c" />, div);
+  expect(div.textContent).to.equal('string, c');
+  await promise;
+  await sleep(1);
+  expect(div.textContent).to.equal('bigint, 123');
+});
+
+test.skip('using isEqual [type checking]', async () => {
+  const use = (...args: unknown[]) => args;
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  function expectType<T>(_expression: T) {
+    // intentionally empty
+  }
+
+  class FakeReplicache<Tx> implements Subscribable<Tx> {
+    subscribe<Data>(
+      query: (tx: Tx) => Promise<Data>,
+      options: {
+        onData: (data: Data) => void;
+        isEqual?: ((a: Data, b: Data) => boolean) | undefined;
+      },
+    ): () => void {
+      use(query, options);
+      return () => undefined;
+    }
+  }
+
+  {
+    const s = useSubscribe(
+      new FakeReplicache<ReadTransaction>(),
+      tx => {
+        use(tx);
+        return Promise.resolve(123n);
+      },
+      {isEqual: (a, b) => a === b},
+    );
+    expectType<bigint | undefined>(s);
+  }
+
+  {
+    // This is invalid because the return type is not json
```

Comment:
I realized this kind of advanced type checking could be put into replicache/reflect too.

I don't think it makes sense to have advanced type check here but not in replicache/reflect.